pkg/syncer/syncer.go

```diff
@@ -67,15 +70,65 @@ const SyncDown SyncDirection = "down"
 // SyncUp indicates a syncer watches resources on the target cluster and applies the status to KCP
 const SyncUp SyncDirection = "up"
 
-func StartSyncer(ctx context.Context, upstream, downstream *rest.Config, resources sets.String, kcpClusterName logicalcluster.LogicalCluster, pcluster string, numSyncerThreads int) error {
-	specSyncer, err := NewSpecSyncer(upstream, downstream, resources.List(), kcpClusterName, pcluster)
+func StartSyncer(
+	ctx context.Context,
+	upstream, downstream *rest.Config,
+	resources sets.String,
+	kcpClusterName logicalcluster.LogicalCluster,
+	pcluster string,
+	numSyncerThreads int,
+	importPollInterval time.Duration,
+) error {
+	// Start api import first because spec and status syncers are blocked by
+	// gvr discovery finding all the configured resource types in the kcp
+	// workspace.
+	apiImporter, err := NewAPIImporter(upstream, downstream, resources.List(), kcpClusterName, pcluster)
+	if err != nil {
+		return err
+	}
+	go apiImporter.Start(ctx, importPollInterval)
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(upstream)
+	if err != nil {
+		return err
+	}
+	fromDiscovery := discoveryClient.WithCluster(kcpClusterName)
+
+	// Block syncer start on gvr discovery completing successfully and
+	// including the resources configured for syncing. The spec and status
+	// syncers depend on the types being present to start their informers.
+	var gvrs []string
+	err = wait.PollImmediateInfinite(gvrQueryInterval, func() (bool, error) {
```

Comment:
Possible follow-up: set a condition on the WorkloadCluster that it's waiting for API import/discovery, for user visibility

Reply:
I'm not sure how critical this is yet - our current model of only syncing a few resource types takes on the order of seconds. Maybe more details status can wait until enough resources are synced that import takes on the order of 10s of seconds?

Reply:
Totally fine!